server-netty/src/main/java/org/jboss/aerogear/webpush/netty/WebPushFrameListener.java

```diff
@@ -133,296 +180,354 @@ public int onDataRead(final ChannelHandlerContext ctx,
 
-    private static Optional<Client> clientForRegId(final String regId) {
-        return Optional.ofNullable(monitoredStreams.get(regId)).orElse(Optional.empty());
-
+    private Http2Headers pushMessageHeaders(final PushMessage pushMessage) {
+        final String pushMessageToken = webpushServer
+                .generateEndpointToken(pushMessage.id(), pushMessage.subscription());
+        return resourceHeaders(Resource.PUSH_MESSAGE, pushMessageToken, EXPOSE_HEADERS_LOCATION);
     }
 
-    private static void logFutureError(final Future future) {
-        if (!future.isSuccess()) {
-            LOGGER.error("ChannelFuture failed. Cause:", future.cause());
+    private static Optional<String> getPushReceiptToken(final Http2Headers headers) {
+        final ByteString byteString = headers.get(PUSH_RECEIPT_HEADER);
+        if (byteString != null) {
+            return extractToken(byteString.toString(), Resource.RECEIPT);
         }
+        return Optional.empty();
     }
 
-    private void handleDeviceRegister(final ChannelHandlerContext ctx, final int streamId) {
-        final Registration registration = webpushServer.register();
-        ctx.attr(REG_ID).set(registration.id());
-        encoder.writeHeaders(ctx, streamId, registrationHeaders(registration), 0, true, ctx.newPromise());
-        LOGGER.info("Registered {}", registration);
-    }
-
-    private static AsciiString asLink(final URI contextUri, final String relationType) {
-        return new AsciiString("<" + contextUri + ">;rel=\"" +  relationType + "\"");
+    private static Optional<Integer> getTtl(final Http2Headers headers) {
+        final ByteString byteString = headers.get(TTL_HEADER);
+        if (byteString != null) {
+            Optional.of(byteString.parseAsciiInt());
+        }
+        return Optional.empty();
     }
 
-    private void handleSubscribe(final ChannelHandlerContext ctx, final String path, final int streamId) {
-        final Optional<Subscription> subscription = extractRegistrationId(path, Resource.SUBSCRIBE.resourceName())
-                .flatMap(webpushServer::newSubscription);
-        subscription.ifPresent(ch -> {
-            LOGGER.info("Subscription {}", ch);
-            notificationStreams.put(ch.endpoint(), ch.registrationId());
-            encoder.writeHeaders(ctx, streamId, createdHeaders(ch), 0, true, ctx.newPromise());
+    private static Http2Headers messageToLarge() {
+        return new DefaultHttp2Headers(false)
+                .status(REQUEST_ENTITY_TOO_LARGE.codeAsText())
+                .set(ACCESS_CONTROL_ALLOW_ORIGIN, ALLOW_ORIGIN_ANY);
+    }
+
+    private void handleReceivingPushMessages(final ChannelHandlerContext ctx,
+                                             final int streamId,
+                                             final Http2Headers headers,
+                                             final String path) {
+        extractToken(path).flatMap(webpushServer::subscriptionById).ifPresent(sub -> {
```

Comment:
@idelpivnitskiy Yeah, there would be no response in that case. But you bring up a valid point the return statuses of the push service, so I'd say either way is fine. We can always revisit later if needed. If you've already added that then lets keep it.